### ISSUE TITLE
Atom 1.6 Frequently Freezes with Remote Project Folders (All Platforms)

### ISSUE THREAD

**Issue by Rondo5000** [atom/atom#11214]
### Description

*Note - Title was edited to reflect that the issue is reported on all platforms, though the following report was originally for Ubuntu 14.04*

Since updating to version 1.6.0 (from 1.5.4, where the problem didn't exist) Atom freezes when trying to open files in networked project folders (i.e. remote folders added to an Atom Project) which are connected via SFTP _(edit: Confirmed to also occur on Samba, SSHFS, etc )._ I have run Atom in safe mode and also tried 1.7 beta and had the same problem.  The networked folders are connected through the built-in "Connect to Server" feature in Ubuntu's Nautilus file browser.  Ubuntu 14.04 was the platform tested.

The **Editor is not responding** message appears after waiting for nearly thirty seconds, and choosing **Keep Waiting** sometimes unfreezes the window, though sometimes not. The problem happens again as soon as another file is selected from the project tree. Atom also commonly freezes when restarting the program with these project folders added, and if it does eventually unfreeze it _will_ soon freeze again when trying to open a project file.

The problem occurs with linked folders on local virtual servers as well as remote servers, so high WAN latency does not appear to be the issue. Atom does not seem to have any CPU spikes during the freeze either as far as I can tell.

### Steps to Reproduce

1. Mount a remote folder via SSH/SFTP in Ubuntu using its Nautilus file browser ('Connect to server' option).
2. Add the folder to an Atom project using the standard folder browser.
3. Once folder is added try to view folder structure or open a file in Atom's tree view.

**Expected behavior:** 
Expected behavior is that the remote project folder operates identically to a local project folder.

**Actual behavior:**
Atom repeatedly freezes for an extended period of time when using the project tree viewer for remote folders and files.

### Versions
Problem occurs in Atom 1.6.0 and later.  Problem did not exist in Atom 1.5.4.  Platform is Ubuntu 14.04 LTS.

**Comment by lhemala** [comment]
I experience the exact same problem on Windows 10.

But instead of using Nautilus (duh), I'm using "SFTP Net Drive Free" to mount a folder of a local virtual machine (Debian Jessie) as a network drive. 

Should I open a separate issue because of the different OS?

**Comment by Sigzegv** [comment]
same issue here on archlinux, mouting folder throught sshfs, it freezes as soon as I try to open a file (but I can browse directories)

**Comment by dborsatto** [comment]
I have a very similar problem, Atom 1.6 randomly freezes when working on files on a Samba share on Windows 10.
I attached a screenshot where you can clearly see that the editor basically stops responsing for about 10 seconds.

![atom-lag](https://cloud.githubusercontent.com/assets/94651/13914872/89065098-ef51-11e5-99bd-d781368d1649.png)

**Comment by dborsatto** [comment]
I traced the bottleneck to 

>relativizeToWorkingDirectory (_path) {
>    return this.getRepo()
>      .then(repo => this.relativize(_path, repo.workdir()))
>}


in src/git-repository-async.js (line ~165), which eventually leads to
>function _tickCallback() {
>...
>}

In node.js, line ~335, whose executions lasts about 10 seconds. It's making the editor unusable, because I work on Samba shares. Is there a way to downgrade to 1.5 while this is taken care of?

**Comment by Rondo5000** [comment]
@dborsatto The download page for all versions of 1.5.4, the latest release without this issue, is at: 
https://github.com/atom/atom/releases/tag/v1.5.4

I cannot speak for the Windows version, but in Ubuntu I was able to uninstall 1.6.0 (_without_ purging settings) and reinstall 1.5.4 without any settings or plugin conflicts occurring, but mileage may vary.

**Comment by tshazar** [comment]
I am having the exact same issue as described here on OSX El Capitan (10.11.3) using a SMB file share as source folder  v.1.5.4 works without issue. Thanks @dborsatto for the v.1.5.4 link.

**Comment by DOSarrest** [comment]
Confirmed, I am using 1.6.0 without any samba or sshfs shares and is doing the same thing. 1.6.0 is currently unusable.

**Comment by JohnRoach** [comment]
I am having the same issue for SSHFS mounted folder. Previous release does not have this issue.

**Comment by sergiouve** [comment]
I can confirm the same issue happening on Fedora 23 while working with SMB.

**Comment by rainbow-pooping-unicorn** [comment]
same problem on mac OS X Yosemite 10.10.5 - gone back to 1.5.4

**Comment by eladts** [comment]
I have the same issue on OS X 10.11.4 working on a remote git project (mounted using SSHFS 2.5.0 and OSXFUSE 2.8.3). This bug makes Atom unusable. If a fix is not available very soon, version 1.6.0 should be pulled.

**Comment by dborsatto** [comment]
@Rondo5000 could you change the title? The issue is not related to SFTP only, from what I can gather it happens with every kind of network directory (SFTP, Samba, SSHFS etc).

I work with a Ubuntu VM guest as a local server, using Windows 10 as a host, and until now I used Samba to share a folder from inside my VM (I work with symlinks so I can't use NTFS) so I could access it from the host. I knew of performance issues with Git, that's why I can't use a GUI (like SourceTree or the recent GitKraken) because of unreasonable loading times, so until now I've stuck with command line via SSH, or using Ungit (which gets executed inside the server).
Now it sounds like whatever Atom was doing, it was more performant under these circumstances. If they won't find a fix, I'll either have to stick with 1.5.4, or I'll have to start using Sublime (please please please don't make me do that).

**Comment by clovis** [comment]
Same issue here on OS X El Capitan using SSHFS and Fuse. As I only work with remote drives, this makes atom unusable...

**Comment by HPieters** [comment]
Exactly the same issue as @clovis... keeps freezing when using remote drives.

**Comment by nicklevett** [comment]
Same as @tshazar but on OS X Yosemite. Downgraded to 1.5.4

**Comment by Knuxchan** [comment]
Same issue here. Downgraded to 1.5.4 and everything works fine again. No freezing/unresponsiveness. I am using Windows 10.

**Comment by CDawg** [comment]
I can reproduce this issue every time and it is has made Atom unusable when trying to open any project using a mounted drive.
To quickly remedy the issue:
1. I dismount the drive
2. Kill Atom & Atom Helper
3. I rename ".git" in my project folder to just "git" so that atom will not read it.
4. Remount the drive
5. Open Atom back up and it seems to be working smoothly.

Somewhere it's reading .git and hanging. (regardless of git version or size of project)

OS: Mac 10.10.5
Atom: 1.6.0
SSHFS version 2.5 (OSXFUSE SSHFS 2.5.0)
OSXFUSE library version: FUSE 2.7.3 / OSXFUSE 2.8.2

**Comment by nathansobo** [comment]
Made some possible progress narrowing this down. I can reliably freeze a vanilla Electron app when reading directory entries from an SSHFS mount when capturing a timeline or profiling. I'm sure many people experiencing this issue were not recording a profile at the time, but I wonder if investigating the cause of that problem might lead to more insight into this problem.

Are people seeing this happen when expanding directories in the tree view? That would match up with where I'm seeing the lock-up occur with profiling enabled in my testing. What *specific* scenarios are causing freezes?

**Comment by dborsatto** [comment]
This happens randomly. Sometimes when navigating the tree view, sometimes saving a file, sometimes when idle.
Looking at the changelog, I honestly thought this problem was connected to the changes made to the underlying git engine, but I suppose I was wrong.

**Comment by nathansobo** [comment]
You might not be wrong, but I'm having a hard time inspecting what's happening due to the dev tools related lock ups. Has anyone had these lock ups in non-git folders?

**Comment by nathansobo** [comment]
@dborsatto How did you form your hypothesis about Git? I'd love to retrace your steps myself.

**Comment by HPieters** [comment]
@nathansobo For me the freezing occurred whilst opening new (previous unopened) files or saving files with changes. 

Downgraded to 1.5.4 and everything is working fine again.

**Comment by nathansobo** [comment]
Is it always happening when interacting with the tree view?

**Comment by dborsatto** [comment]
@nathansobo Well, I looked at the changelog and I saw this implementation switch: https://github.com/atom/atom/pull/9213
I've know for a while of git performance issues while working with remote project folders, so I made an educated guess, also confirmed by what @CDawg said on https://github.com/atom/atom/issues/11214#issuecomment-200474938
Of all the stuff that changed in 1.6, this seemed to me the one that most likely could be the culprit. Unless the problem is with a dependency, such as Electron itself, of course.

**Comment by HPieters** [comment]
@nathansobo Just updated to 1.6.0 to check, it does not freeze on interaction with tree view. The only consistent freeze is when saving, most of the time.

Interesting behaviour is that I do get a freeze every time atom goes to the background (e.g. go to chrome to type this message) when I go back to atom it freezes on using the tree view.

**Comment by nathansobo** [comment]
So, with the help of @as-cii I'm pretty sure we've found the culprit. We took a process profile during a slow interaction and discovered that the majority of the time was being spent in node-git checking for ignored files. Our node-git-based `AsyncRepository` implementation is new as of 1.6 so this makes sense. I think our old implementation probably performed vastly fewer file system calls.

![screen shot 2016-03-24 at 5 05 45 am](https://cloud.githubusercontent.com/assets/1789/14015458/c59db12a-f17e-11e5-87df-da53a1ce13ba.png)

[Sample of Atom Helper.txt](https://github.com/atom/atom/files/187683/Sample.of.Atom.Helper.txt)

@joshaber could you take a look at the code paths involved here and see if we might apply some caching or some other technique to reduce the thrashing here?

**Comment by Knuxchan** [comment]
@nathansobo This problem occurred for me when I tried interacting with the tree view (like clicking on files/folders). Also, if I try to 'Add Project Folder' to view in the tree view, Atom will freeze up.

**Comment by joshaber** [comment]
@nathansobo Ha, so that stack trace is actually `git-utils`, not async git. So that hasn't changed. It doesn't look terribly suspicious. Just the callstack for `git_ignore_path_is_ignored`.

One theory I had is that since async git is ... async ... we could be hitting the FS in parallel a lot more than we did before, leading to more disk contention, which network file systems might not handle so well.

**Comment by guybrush** [comment]
i have the same issue i think: 5-10sec freeze everytime i refocus on the atom-window

* im on windows 10
* my projects are mounted via samba (from virtualbox linux-guest)
* i dont have the problem with 1.5.3

**Comment by adamfranco** [comment]
Just wanted to note that this seems to be a very similar issue (duplicate?) of #1766 . 

In that issue [I confirmed that](https://github.com/atom/atom/issues/1766#issuecomment-201344724) Atom's Git implementation is scanning and reading every file in the project (using ~26 `lstat` and `open` calls for each file) and doing this every time Atom is brought to the foreground, a process that may take many minutes depending on the network speed and project size. Additionally, these check cycles seem to get queued up and my run one after another, leading to even longer periods of contention.

**Comment by nathansobo** [comment]
We're getting traction on this, but anyone using network mounts should fall back to 1.5.x and disable automatic updates until this is resolved.

**Comment by dborsatto** [comment]
@adamfranco woah, that certainly is a lot of calls! Why so many?

**Comment by Maxr1998** [comment]
I also have this issue, but I'm still on 1.2.4.. I wonder why auto-update didn't work. :P
Anyway, I'll upgrade to 1.5.4 until the issue is resolved.

**Comment by adamfranco** [comment]
@dborsatto For reference, [this gist shows all 388 `lstat` and `open` calls](https://gist.github.com/adamfranco/804c161bfbb2f951ce4a) made when bringing Atom into focus with my small test Project Folder (3 files).

Looking at this, I notice that the project directory, `.git` directory, and some `.git` subdirectories get scanned sometimes several times during the status check for each file. It may be that the Git implementation is searching the the `.git` sub-directories for objects and not caching objects or the directory listings.

**Comment by nathansobo** [comment]
Trouble is, if you cache, when do you bust that cache? Simple timeout?

**Comment by adamfranco** [comment]
@nathansobo, I wonder if it would be reasonable the Git implementation to consider the repository contents static/atomic for the length of the entire "check the git status for the Project Folder" operation. If `.git/HEAD` is the first thing loaded, all of the objects referenced by that ref should already exist on the file-system and (if I remember how git works correctly, which is debatable :wink:) should persist without modification until `git gc` is run. If other git operations add new objects/refs/tags/heads/etc, those referenced by the initial look at `.git/HEAD` would still be there. I would think that to make such a system work the caller would need to pass either some sort of cache object or cache-id to the git implementation to use for each of the `git status` checks.

**Comment by joshaber** [comment]
1.6.1 is now out which should make life more bearable. It's still not as fast as 1.5.4 was. We're continuing to work on it.

**Comment by dborsatto** [comment]
@joshaber Thanks for the work! I love what you're all doing.

**Comment by joshaber** [comment]
As of #11315 performance should be back on par with 1.5.4. It'll be released publicly in 1.8.

I'm sorry this painful regression happened. Thanks for all the reports!

**Comment by dborsatto** [comment]
@joshaber how come Atom 1.8 and not 1.7? 1.6.1 has definitely made some great progress (freezes for me occur very rarely) and I can use it rather painlessly, but I have a local network share (from the VM inside my PC) on an SSD, so I'm kind of lucky, and it's weird to think that people on really slow networks need to wait 2 minor releases  before upgrading from 1.5.
I'm not sure what the patch policy is, but seeing this is critical and blocking for some users, couldn't the commits be cherry-picked into 1.7?

**Comment by joshaber** [comment]
That's a fair question @dborsatto. Because they are fairly large changes, I'd like them to go through the standard beta testing process so that we can try to catch any new bugs before it's released.

**Comment by Rondo5000** [comment]
I have only today been able to test the 1.6.1 fix (I created the original report) and Atom remains unusable for me because of problems that _may_ have originated from the changes made to remedy the freezes.  I am now experiencing the blank files issue from #11327 instead of the lock-ups.  The remote files which I would be attempting to open only appear after a long time which is comparable to the extended freeze I reported above.  So essentially, Atom remains in a responsive state instead of freezing, but the desired action of the remote file contents actually appearing for editing remains just as delayed as prior to these changes.

I'll note that as in the original report I am still using Ubuntu 14.04 and an SSH mounted drive, and I'll _also_ note that I'm horribly sorry to keep being the bearer of bad news(!), and my thanks to all others involved.

**Comment by eddy147** [comment]
@joshaber Fair enough. But for people like me, who on working on big remote projects, I have to wait until 1.8 is out.

**Comment by joshaber** [comment]
@eddy147 Alternatively, you could download 1.5.4 and disable updates until 1.8 is out.

As for timing, no guarantees, but I'd expect 1.7 to be out in a week or two, which means 1.8 would be available for download as a beta then. And then another few weeks until 1.8 is out as the stable release.

**Comment by eddy147** [comment]
@joshaber Thank you!

**Comment by Arcanemagus** [comment]
Btw, v1.8.0-beta0 is available as of ~2 hours ago :wink:.

**Comment by bakkerthehacker** [comment]
Still an issue on 1.8.0 linux mint 17.3.  Alt-tabbing freezes the window for several seconds
1.5.4 is working fine

**Comment by Enelar** [comment]
bump for reopening. @joshaber 
i can't type, while working on `gvfs-sftp`. it just hangs for 2-3 seconds during typing.

**Comment by dsalex** [comment]
1.12.5 on Ubuntu still freezes all the time, almost unusable with low ping (through VPN or something).

**Comment by Enelar** [comment]
I noticed, that it freezing each time it check is file changed on disk. Look like they did it synchronous.

**Comment by kylefarris** [comment]
Still having the same issue in the latest version. Actually, it was working really well until I did a large git commit on my repository and now Atom just times out over and over again.

**Comment by Enelar** [comment]
@kylefarris i managed to workaround it by mounting with cache timeout 1 hour and more. As far as i understand, the issue is caused by underlying architecture: eveerything requests in man thread, synchronymous mode. You may see, when you're reloading folder content, whole UI is freezing.
Same thing with git plugin, file saving etc.
For example you cant switch tab while file save in process. 

Try to disable atomic save, and increase sysctl vm timeouts.

**Comment by kylefarris** [comment]
@Enelar Thanks! I'm generally experiencing this when I'm working on my Windows 10 machine at work unfortunately so the sysctl part won't help much but atomic save may. Is the atomic save a config option in Atom itself?
